### PR TITLE
PB-494: Refactor to make ready for Woocommerce

### PR DESCRIPTION
**Pull request by peteward**
* Remove models
* separate comms or server errors from request content errors (so we don't retry errors for states that won't change) 
* introduce a base exception that client apps can catch for all PB errors

Once this is merged I intend to:
* create a new version of this under `php-sdk`
* remove the `-NEW` endpoint after switching over the server-side code
* implement the print endpoint
* flesh out the documentation

### CODE REVIEW

**Comment by peteward** [comment]
Reasons for removing models:

1). It enforces a cumbersome workflow on our side when we need to make changes to our API. This could become more frequent as this starts to be used - new platforms, new customers, new data, new use-cases.

Currently you have to:
* Make the required change server-side (if necessary)
* Deploy
* Make corresponding changes here with tests
* Deploy
* Update each individual client to update the dependency to the latest
* Deploy changes

Without enforcing models we can cut out any changes to this package just because we made a change to our API endpoint and iterate a lot faster.
In the case of people using our package for bespoke platforms we only need to make changes server-side.

2). I would argue that current DX wasn't great:

```
        return PennyBlackOrder::fromValues(
            $order->getId(),
            $order->getIncrementId(),
            $order->getCreatedAt(),
            $this->customerMapper->map($order),
            $this->orderDetailsMapper->map($order)
        );
```

vs constructing as an array that matches the endpoint definition in stoplight (which I personally find a lot clearer):

```
        return [
            "customer" => [
                "vendor_customer_id" => (int) $order->get_customer_id(),
                "first_name" =>  $order->get_shipping_first_name() ?: ($user ? $user->first_name : ''),
                "last_name" => $order->get_shipping_last_name() ?: ($user ? $user->last_name : ''),
                "email" => $order->get_billing_email(),
                "language" => $this->mapLocaleToLanguage($user->get_locale()),
                "marketing_consent" => null, // TODO: Unknown by default, needs a plugin
                "total_orders" => count($allCustomerOrders),
                "tags" => [], // TODO: Tagging isn't native, it's provided by plugins
                "total_spent" => array_sum(array_map(function ($order) {
                    return $order->get_total();
                }, $allCustomerOrders)),
            ],
            "order" => [
                "id" => (string) $order->get_id(),
                "number" => $order->get_order_number(),
                "created_at" => $order->get_date_created() ? $order->get_date_created()->format('Y-m-d H:i:s') : date('Y-m-d H:i:s'),
                "total_amount" => (float) $order->get_total(),
                "total_items" => (int) $order->get_item_count(),
                "billing_country" => (string) $order->get_billing_country(),
                "billing_postcode" => (string) $order->get_billing_postcode(),
                "billing_city" => (string) $order->get_billing_city(),
                "shipping_country" => (string) $order->get_shipping_country(),
                "shipping_postcode" => (string) $order->get_shipping_postcode(),
                "shipping_city" => (string) $order->get_shipping_city(),
                "currency" => (string) $order->get_currency(),
                "gift_message" => '', // TODO: Gift message isn't native, it's provided by plugins.
                "skus" => $skus,
                "product_titles" => $titles,
                "promo_codes" => $order->get_coupon_codes(),
                "is_subscription_reorder" => false,
            ],
            "origin" => self::ORIGIN_WOOCOMMERCE,
        ];
```

Whilst the models were really useful when previously there was no feedback to the success of the request (it was passed to event bus so very much 'hit and hope'), we can now get good feedback on what is wrong to help guide the implementation - much better DX and less need for the strictness which I think just gets in the way.

Admittedly, the above could have DX improved with method chaining like:
```
$order = new PBOrder();
$order->setId(1)
    ->setNumber("10001")
    ->setCreatedAt(X);
   ...

```

but we still have point 1). above which seems sensible to me for a fast-moving start-up.

If you feel very strongly that we should keep the models then we could add in methods like above to keep DX and make the sending method available to either an array or a model to give more flexibility

wdyt?

**Comment by jameshalsall** [comment]
I see what you mean, but I still feel like "give me an array" is less helpful than offering setters, I agree the `fromValues()` is not that nice DX either **but** at least you get the parameter names which hint at what you can provide. My problem with arrays is lack of typing + value hinting. 

To counter point 1, in an ideal world we should have some shared definition of the API and the SDKs (in various languages) should be generated from that ideally, up to a point. This is what a gRPC + protobuf workflow is like, and you don't write any of the SDK yourself. We are a way off having something like that though.

I'm on board with removing these for now because, as you mentioned, they will speed things up in the short term, but I still feel like the DX is worse by asking for an array of values than offering helpful setters. This can be something we look at in the future though if we have more external developers using our SDK 👍🏻

**Comment by peteward** [comment]
ok let's compromise.
I'll revert, keep the models and improve the DX 👍

File: example/InstallExample.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+$httpClient = new Client();
+$streamFactory = new HttpFactory();
+$requestFactory = new HttpFactory();
+
+$apiKey = "YOUR-API-KEY";
+$isTest = true;
+
+$api = new Api($httpClient, $requestFactory, $streamFactory, $apiKey, $isTest);
+
+$api->install("store.pennyblack.io");
```
Comment (jameshalsall) [inline example/InstallExample.php:18]:
```suggestion
$api->install("mystore.example.com");
```

for complete clarity that this isn't some fixed Penny Black domain that would be used

File: src/Model/Customer.php [hunk 2]
```diff
@@ -2,90 +2,93 @@
...
+    }
+
+    public function setTags(array $tags): Customer
+    {
         $this->tags = $tags;
+
+        return $this;
+    }
+
+    public function setTotalSpent(float $totalSpent): Customer
```
Comment (jameshalsall) [inline src/Model/Customer.php:73]: you can use `): self` as the return type I think, and then if it was ever extended the typehint remains correct (returns `$this` which may be something that extends `Customer`)

File: src/Model/Customer.php [hunk 3]
```diff
@@ -2,90 +2,93 @@
...
+        return [
+            "vendor_customer_id" => $this->vendorCustomerId ?? null,
+            "first_name" => $this->firstName,
+            "last_name" => $this->lastName,
+            "email" => $this->email,
+            "language" => $this->language ?? null,
+            "marketing_consent" => $this->marketingConsent ?? null,
+            "total_orders" => $this->totalOrders ?? null,
+            "tags" => $this->tags,
+            "total_spent" => $this->totalSpent ?? null,
```
Comment (jameshalsall) [inline src/Model/Customer.php:91]: wouldn't this default to 0?

File: src/Model/Customer.php [hunk 4]
```diff
@@ -2,90 +2,93 @@
...
+    public function toArray(): array
+    {
+        return [
+            "vendor_customer_id" => $this->vendorCustomerId ?? null,
+            "first_name" => $this->firstName,
+            "last_name" => $this->lastName,
+            "email" => $this->email,
+            "language" => $this->language ?? null,
+            "marketing_consent" => $this->marketingConsent ?? null,
+            "total_orders" => $this->totalOrders ?? null,
```
Comment (jameshalsall) [inline src/Model/Customer.php:89]: wouldn't this default to 0?

File: src/Api.php [hunk 5]
```diff
@@ -2,50 +2,89 @@
...
 
     /**
-     * @throws Exception\AuthenticationException
-     * @throws RequestException
+     * @throws PennyBlackException
      */
     public function install(string $shopUrl): void
     {
-        $this->client->jsonRequest('POST', 'install', ['shop_url' => $shopUrl]);
+        $this->sendPostRequest('ingest/install', ['shop_url' => $shopUrl]);
```
Comment (jameshalsall) [inline src/Api.php:50]: interesting, if this SDK was ever to be widened to beyond just ingest stuff it might be worth us changing this function name to `ingestInstall()` to reflect the path here, **but** that is a bit weird to the developer 🤔 

maybe leave as is for now, I can't see us having any other kind of install in the forseeable anyway

File: src/Model/Customer.php (hunk 3 above)
Comment (peteward) [reply src/Model/Customer.php:91]: I think null is better for those who don't have the information available, rather than just defaulting to 0. Although it would probably end up as the same thing on our end 🤔

File: src/Model/Customer.php (hunk 4 above)
Comment (peteward) [reply src/Model/Customer.php:89]: as above

File: src/Api.php (hunk 5 above)
Comment (peteward) [reply src/Api.php:50]: maybe I will clarify as `installStore`

File: src/Model/Customer.php (hunk 3 above)
Comment (peteward) [reply src/Model/Customer.php:91]: we could make it optional on our API, so we don't send it if we don't know it, as opposed to allowing it to be nullable.

I've just changed the server-side to have all values as required and allow them to be nullable. I was on the fence about this though, so happy to change approach and only include them if they're set if you prefer

Comment (jameshalsall) [reply src/Model/Customer.php:91]: on your example as well, a total_orders of zero would be fine, because we know there is no customer because the ID is null 👍🏻

Comment (jameshalsall) [reply src/Model/Customer.php:91]: actually quite odd that the marketing consent is true there, when it should be null?

Comment (peteward) [reply src/Model/Customer.php:91]: > optional being there is some default when it's not provided (which may or may not be a null value)

I don't agree with this part. I think if something is optional, then we don't record it and store it at all.

Currently, we do record and store lots of values as null, but I don't think this is scalable and sensible for the longer term.  We will have more platforms, more customers, more use-cases leading to more fields and values and the use of them will need to be dynamic to some degree. 

I'm thinking that we should have our minimal set of core mandatory fields, and then start to make many of the other fields optional. We will store what we have and segment / render what is available. This will reduce our data storage footprint as we're not storing null / empty / zero values for 75% of fields because the merchant can only provide a subset - we store what we know.

At the moment we fail a match if a value is null for rendering, so this part is already in place, but we would also need to gracefully handle failing when an attribute/key does not exist. Further down the line we can make the availability of fields for rendering and segmenting dynamic by examining the order data being provided and enabling/disabling pbvs based on what they've got.

This ties in with [this ADR discussion](https://www.notion.so/pennyblack/Capture-groupings-of-orders-via-our-ingest-API-d8cbdf3f4f8841598f327a01ae1e069e?pvs=4).  I certainly don't want to add an explicit field that all clients need to add and set as null if not used on every platform. This `retailer`, or whatever we call it, would be an optional value. 

In summary I think I should change these values that do not necessarily need to be mandatory, make them optional in our server-side validation, only send them here if they're set, and allow optional parameters to work throughout our platform.

Comment (jameshalsall) [reply src/Model/Customer.php:91]: Yeah I agree with this 💯 - when I've been talking about "optional values having defaults" I'm talking in the grand scheme of the platform, so if something is not provided on the order then yeah we don't store anything in the JSON schema for it as it's wasteful, but when we evaluate that data if something is not there because it's an optional field then we may well have a default for it

e.g. the omission of `total_spent` would indicate 0

Comment (peteward) [reply src/Model/Customer.php:91]: ok, I'll get the server-side changes done to support this shift

**Review by jbarton123 — COMMENTED** [review]
Very nice 👍🏻
